src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -3104,11 +3108,215 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                                                                                   baseArgument.InConversionConvertibleOpt,
+                                                                                   baseArgument.OutConversionConvertibleOpt,
+                                                                                   semanticModel: null, arg.Syntax, arg.IsImplicit));
+                            }
+
+                            propertyArguments = propertyArgumentsBuilder.ToImmutableAndFree();
+                        }
+
+                        Debug.Assert((propertyReference == null && propertyArguments.IsEmpty) ||
+                                     (propertyArguments.Length == propertyReference.Arguments.Length));
+
+                        // If there is control flow in the value being assigned, we want to make sure that
+                        // the instance is captured appropriately, but the setter/field load in the reference will only be evaluated after
+                        // the value has been evaluated. So we assemble the reference after visiting the value.
+
+                        _evalStack.Push(Visit(memberReference.Instance));
+                        right = Visit(assignmentOperation.Value);
+
+                        IOperation instance = _evalStack.Pop();
+                        switch (memberReference.Kind)
+                        {
+                            case OperationKind.FieldReference:
+                                var fieldReference = (IFieldReferenceOperation)memberReference;
+                                left = new FieldReferenceExpression(fieldReference.Field, fieldReference.IsDeclaration, instance, semanticModel: null,
+                                                                    fieldReference.Syntax, fieldReference.Type, fieldReference.ConstantValue, fieldReference.IsImplicit);
+                                break;
+                            case OperationKind.EventReference:
+                                var eventReference = (IEventReferenceOperation)memberReference;
+                                left = new EventReferenceExpression(eventReference.Event, instance, semanticModel: null, eventReference.Syntax,
+                                                                    eventReference.Type, eventReference.ConstantValue, eventReference.IsImplicit);
+                                break;
+                            case OperationKind.PropertyReference:
+                                left = new PropertyReferenceExpression(propertyReference.Property, instance, propertyArguments, semanticModel: null, memberReference.Syntax,
+                                                                       memberReference.Type, memberReference.ConstantValue, memberReference.IsImplicit);
+                                break;
+                        }
```

Comment:
>} [](start = 24, length = 1)

Do we need a `default` case here?

Reply:
No, we're already inside a switch restricting the kind to these three choices.

---
In reply to: [185550604](https://github.com/dotnet/roslyn/pull/26389#discussion_r185550604) [](ancestors = 185550604)